Fixes #22

sarah-fidahussain-SLU opened the pull request:
Fixes https://github.com/oss-slu/TPEN-interfaces/issues/22

What was changed: I added two functions: one function returns a user's highest privilege role if the user has many roles and also checks if the user has invalid roles. Another function checks if the user has permission to perform a certain task based on their highest privilege role.
Why it was changed: We needed to implement logic to handle multiple roles for a single user while also ensuring the highest privilege role is prioritized. We also needed to ensure that only certain roles can access sensitive information.
How was it changed: In the main.js, I added a getHighestPrivilegeRole() and userHasAccess(), and throughout the functions, they log errors in the console if something is invalid.

camilledaugherty commented on components/classroom/main.js:
```diff
@@ -42,3 +45,50 @@ window.onload = () => {
+const ROLE_HIERARCHY = [Roles.OWNER, Roles.LEADER, Roles.CONTRIBUTOR];
+
+// identifies the highest privilege role from a list of user roles and returns the top-ranked valid role
+function getHighestPrivilegeRole(userRoles) {
```
Maybe in future sprints we have userRoles as an attribute of a User object, just so we start to see some connection between all of these structures!

MutyaSowmya123 replied:
Yes we will be working on them shortly and to confirm I just want to clarify the role attached to a user it may be any of the 3 roles we have

camilledaugherty commented on components/classroom/main.js:
```diff
@@ -42,3 +45,50 @@ window.onload = () => {
+    }
+
+    // Log unexpected roles if any were found
+    if (invalidRoles.length > 0) {
```
Not sure there's a need for an invalidRoles array? If role is not within hierarchy just throw error

camilledaugherty commented on components/classroom/main.js:
```diff
@@ -42,3 +45,50 @@ window.onload = () => {
+}
+
+// checks if a user, based on their highest privilege role, has permission to perform a specified action
+function userHasAccess(userRoles, action, scope, entity) {
```
maybe this can be implemented within hasPermission

MutyaSowmya123 reviewed (changes requested):
I want to add some more clarity on the things so you get to know the changes required.

For the first case where there are multiple roles assigned for a single user,highest priority role could be prioritized but also ensure if any role allows the behavior the action is allowed.

This user will not be able to retrieve the Project from the API and will receive 403 responses on any call to a Project/Line/Page/Group they do not have a role in (unless there is a public user access set). The message returned is usually something like "You are not a member of this Project."

In the back end, unrecognized roles or permissions are discarded. In the front-end, the Project class will retain any unknown roles and the interface logic can ignore/filter/remove them as they like. The back-end should enforce that no Project can have 0 "OWNER" or 0 contributors, so the interface can expect an error if that change is attempted.

sarah-fidahussain-SLU commented:
I just pushed changes concerning front end, I just changed the error message from "project not found" to "Project not found or access denied" and I changed error messaged to 403 errors, and I also made sure that not only the highest role is prioritized, but the user is given access with all the roles given.
